src/ucp/tag/tag_send.c

```diff
@@ -99,23 +109,32 @@ ucp_tag_send_req(ucp_request_t *req, size_t dt_count,
 
     /*
      * Start the request.
-     * If it is completed immediately, release the request and return the status.
+     * If it is completed immediately and this completion is allowed,
+     * release the request and return the status.
      * Otherwise, return the request.
      */
     status = ucp_request_send(req, 0);
     if (req->flags & UCP_REQUEST_FLAG_COMPLETED) {
-        ucs_trace_req("releasing send request %p, returning status %s", req,
-                      ucs_status_string(status));
-        if (enable_zcopy) {
-            ucp_request_put(req);
+        if (!(param->op_attr_mask & UCP_OP_ATTR_FLAG_NO_IMM_CMPL)) {
+            /*  immediate completion is allowed */
+            ucs_trace_req("releasing send request %p, returning status %s", req,
+                          ucs_status_string(status));
+            ucp_request_put_param(param, req);
+            return UCS_STATUS_PTR(status);
         }
-        return UCS_STATUS_PTR(status);
+
+        ucs_trace_req("request %p completed, but immediate completion is "
+                      "prohibited, status %s", req,
+                      ucs_status_string(status));
+        ucp_request_cb_param(param, req, send);
```

Comment:
this seems similar to tag_recv.c lines 70-80, can we unite?

Reply:
unified